htdoc/htdump.1.in
.TH htdump 1 "15 October 2001"
.\" NAME should be all caps, SECTION should be 1-8, maybe w/ subsection
.\" other parms are allowed: see man(7), man(1)
.SH NAME
htdump \- write out an ASCII-text version of the document database
.SH SYNOPSIS
.B htdump
.I "[options]"
.SH "DESCRIPTION"
Htdump writes out an ASCII-text version of the document
database in the same form as the -t option of htdig.
.SH OPTIONS
.TP
.B \-a
Use alternate work files. Tells htdump to append 
.I .work
to database files, allowing it to
operate on a second set of databases.
.TP
.B \-c \fIconfigfile\fR
Use the specified
.I configfile
instead of the default.
.TP
.B \-v
Verbose mode. This doesn't have much effect.
.SH File Formats
.TP
.B Document Database
Each line in the file starts with the document id 
followed by a list of
.I fieldname
:
.I value
separated by tabs. The fields always appear in the
order listed below:
.TP
.B u
URL
.TP
.B t
Title
.TP
.B a
State (0 = normal, 1 = not found, 2 = not indexed, 3 = obsolete)
.TP
.B m
Last modification time as reported by the server
.TP
.B s
Size in bytes
.TP
.B H
Excerpt
.TP
.B h
Meta description
.TP
.B l
Time of last retrieval
.TP
.B L
Count of the links in the document (outgoing links)
.TP
.B b
Count of the links to the document (incoming links or backlinks)
.TP
.B c
HopCount of this document
.TP
.B g
Signature of the document used for duplicate-detection
.TP
.B e
E-mail address to use for a notification message from htnotify
.TP
.B n
Date to send out a notification e-mail message
.TP
.B S
Subject for a notification e-mail message
.TP
.B d
The text of links pointing to this document. (e.g. <a href="docURL">description</a>)
.TP
.B A
Anchors in the document (i.e. <A NAME=...)
.TP
.B Word Database
While htdump and htload don't deal with the word database
directly, it's worth mentioning it here because you need to
deal with it when copying the ASCII databases from one system
to another. The initial word database produced by htdig is
already in ASCII format, and a binary version of it is
produced by htmerge, for use by htsearch. So, when you copy
over the ASCII version of the document database produced by
htdump, you need to copy over the wordlist as well, then
run htload to make the binary document database on the target
system, followed by running htmerge to make the word index.
.TP
Each line in the word list file starts with the word
followed by a list of
.I fieldname
:
.I value
separated by tabs. The fields always appear in the
order listed below, with the last two being optional:
.TP
.B i
Document ID
.TP
.B l
Location of word in document (1 to 1000)
.TP
.B w
Weight of word based on scoring factors
.TP
.B c
Count of word's appearances in document, if more than 1
.TP
.B a
Anchor number if word occurred after a named anchor
.SH "FILES"
.TP
.B @DEFAULT_CONFIG_FILE@
The default configuration file.
.TP
.B @DATABASE_DIR@/db.docs
The default ASCII document database file.
.TP
.B @DATABASE_DIR@/db.wordlist
The default ASCII word database file.
.SH "SEE ALSO"
Please refer to the HTML pages (in the htdig-doc package)
.B (THIS MUST BE CUSTOMISED...)
.B /usr/share/doc/htdig-doc/html/index.html
and the manual pages
.I htdig(1)
,
and
.I htload(1)
for a detailed description of ht://Dig and its commands.
.SH AUTHOR
This manual page was written by Stijn de Bekker, based on the HTML
documentation of ht://Dig.


ChangeLog
Wed Feb 23 16:53:40 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htcommon/defaults.cc (backlink_factor, external_parser,
	local_default_doc, local_urls, local_urls_only, local_user_urls):
	Add some updates from 3.1.5's attrs.html.
	* htdoc/attrs.html, htdoc/cf_byname.html, htdoc/cf_byprog.html:
	Regenerate using cf_generate.pl.

Wed Feb 23 15:11:51 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	[ Improve htsearch's HTML 4.0 compliance ]
	* htsearch/TemplateList.cc (createFromString): Use file name rather
	than internal name to select builtin-* templates, use $&(TITLE) and
	$&(URL) in templates and quote HTML tag parameters.
	* installdir/long.html, installdir/short.html: Use $&(TITLE) and
	$&(URL) in templates and quote HTML tag parameters.
	* htsearch/Display.cc (setVariables): quote all HTML tag parameters
	in generated select lists.
	* installdir/footer.html, installdir/header.html,
	installdir/nomatch.html, installdir/search.html,
	installdir/syntax.html, installdir/wrapper.html:
	Use $&(var) where appropriate, and quote HTML tag parameters.
	* installdir/htdig.conf: quote all HTML tag parameters.

Wed Feb 23 13:40:27 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htcommon/URL.h (encodeURL): Change list of valid characters to
	include only unreserved ones.
	* htcommon/cgi.cc (init): Allow "&" and ";" as input param. separators.
	* htsearch/Display.cc (createURL): Encode each parameter separately,
	using new unreserved list, before piecing together query string, to
	allow characters like "?=&" within parameters to be encoded.

Wed Feb 23 13:22:29 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htlib/URL.cc (ServerAlias): Fix server_aliases processing to prevent
	infinite loop (as for local_urls in PR#688).

Wed Feb 23 12:49:52 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htlib/HtDateTime.h, htlib/HtDateTime.cc: change Httimegm() method
	to HtTimeGM(), to avoid conflict with Httimegm() C function, so we
	don't need "::" override, for Mac OS X.
	* htlib/htString.h, htlib/String.cc: change write() method to
	Write(), to avoid conflict with write() function, so we don't need
	"::" override, for Mac OS X.

Wed Feb 23 12:17:46 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htlib/Configuration.cc(Read): Fixed to allow final line without
	terminating newline character, rather than ignoring it.

Wed Feb 23 12:01:01 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htdig/Retriever.cc (GetLocal, GetLocalUser): Add URL-decoding
	enhancements to local_urls, local_default_urls & local_default_doc,
	to allow hex encoding of special characters.

Wed Feb 23 19:14:29 2000  Loic Dachary  <loic@ceic.com>

 	* htcommon/conf_parser.cxx: regenerated from conf_parser.yxx

Wed Feb 23 19:04:16 2000  Loic Dachary  <loic@ceic.com>

	* test/test_functions.in: inconditionaly remove existing test/var
	  directory before runing tests to prevent accidents.

	* htcommon/URL.cc (URL): fixed String->char warning

	* htcommon/defaults.cc (wordlist_compress): defaults to true

Tue Feb 22 17:09:10 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htdig/HTML.cc(parse, do_tag): Fix handling of <img alt=...> text
	and parsing of words in meta tags, to to proper word separation.
	* htlib/HtWordType.h, htlib/HtWordType.cc: Add HtWordToken() function,
	to replace strtok() in HTML parser.

Tue Feb 22 16:21:25 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htlib/URL.cc (ctor, normalizePath): Fix PR#779, to handle relative
	URLs correctly when there's a trailing ".." or leading "//".

Tue Feb 22 14:09:26 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htdig/Document.cc (RetrieveLocal): Handle common extensions for
	text/plain, application/pdf & application/postscript.

Mon Feb 21 17:25:21 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* contrib/htdig-3.2.0.spec: Fixed %post script to add more
	descriptive entries in htdig.conf, made cron script a config file,
	updated to 3.2.0b2.

	* contrib/conv_doc.pl, contrib/parse_doc.pl: Added comments to show
	Warren Jones's updates in change history.

Mon Feb 21 17:09:13 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htcommon/HtConfiguration.h, htcommon/conf_parser.yxx,
	htlib/Configuration.h, htlib/Configuration.cc: split Add() method
	into Add() and AddParsed(), so that only config attributes get parsed.
	Use AddParsed() only in Read() and Defaults().

Sat Feb 19 04:56:47 2000  Geoff Hutchison  <ghutchis@wso.williams.edu>

	* htnet/Connection.h, htnet/Connection.cc: Renamed methods with
	capitals to remove the need to use ::-escaped library calls.

Fri Feb 18 14:40:50 2000  Loic Dachary  <loic@ceic.com>

	* test/conf/access.conf.in: removed cookies. Not used and some
	  httpd are not compiled with usertrack.

Wed Feb 16 12:15:08 2000  Vadim Chekan  <vadim@etc.lviv.ua>

	* htcommon/Makefile.am replaced conf.tab.cc.h by conf_parser.h in
	noinst_HEADERS

	* htcommon/conf_parser.yxx,conf_parser.lxx,HtConfiguration.cc,
	HtConfiguration.h: added copyright and Id:

	* htcommon/cgi.cc(init): fixed bug: array must be free by 
	delete [] buf, not just delete buf;

Tue Feb 15 23:16:14 2000  Geoff Hutchison  <ghutchis@wso.williams.edu>

	* htnet/HtHTTP.cc (isParsable): Remove application/pdf as a
	default type--it is now handled through the ExternalParser
	interface if at all.

	* htcommon/defaults.cc: Remove pdf_parser attribute.

	* htdoc/attrs.html, htdoc/cf_byname.html, htdoc/cf_byprog.html:
	Regenerate using cf_generate.pl.

	* htdig/Document.cc (getParsable): Remove PDF once and for all
	(hopefully).

	* htdig/ExternalParser.cc (parse): Ditto.

	* configure.in: Remove check for PDF_PARSER.

	* configure: Regenerate using autoconf

	* htdig/Makefile.am: Remove PDF.cc and PDF.h.

	* Makefile.in, */Makefile.in: Regenerate using automake --foreign

Tue Feb 15 12:02:39 EET 2000  Vadim Chekan  <vadim@etc.lviv.ua>

	* htcommon/HtConfiguration.cc,HtConfiguration.h: fixed bug discovered

src/tools/htdump.cc
//
// htdump.cc
//
// htdump: A utility to create ASCII text versions of the document
// and/or word databases. These can be used by external programs,
// edited, or used as a platform and version-independent form of the DB.
//
// Part of the ht://Dig package   <http://www.htdig.org/>
// Copyright (c) 1999-2004 The ht://Dig Group
// For copyright details, see the file COPYING in your distribution
// or the GNU Library General Public License (LGPL) version 2 or later
// <http://www.gnu.org/copyleft/lgpl.html>
//
// $Id: htdump.cc,v 1.1.2.1 2006/09/25 23:51:20 aarnone Exp $
//
#ifdef HAVE_CONFIG_H
#include "htconfig.h"
#endif /* HAVE_CONFIG_H */

#include "WordContext.h"
#include "HtURLCodec.h"
#include "HtWordList.h"
#include "HtConfiguration.h"
#include "DocumentDB.h"
#include "defaults.h"

#include <errno.h>

#ifndef _MSC_VER /* _WIN32 */
#include <unistd.h>
#endif

// If we have this, we probably want it.
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#elif HAVE_GETOPT_LOCAL
#include <getopt_local.h>
#endif

int		verbose = 0;

void usage();
void reportError(char *msg);

//*****************************************************************************
// int main(int ac, char **av)
//
int main(int ac, char **av)
{
    int			do_words = 1;
    int			do_docs = 1;
    int			alt_work_area = 0;
    String		configfile = DEFAULT_CONFIG_FILE;
    int			c;
    extern char		*optarg;

    while ((c = getopt(ac, av, "vdwc:a")) != -1)
    {
	switch (c)
	{
	    case 'c':
		configfile = optarg;
		break;
	    case 'v':
		verbose++;
		break;
	    case 'a':
		alt_work_area++;
		break;
	    case 'w':
	        do_words = 0;
	        break;
	    case 'd':
	        do_docs = 0;
	        break;
	    case '?':
		usage();
		break;
	}
    }

   HtConfiguration* config= HtConfiguration::config();
   config->Defaults(&defaults[0]);

    if (access((char*)configfile, R_OK) < 0)
    {
	reportError(form("Unable to find configuration file '%s'",
			 configfile.get()));
    }
	
   config->Read(configfile);

    //
    // Check url_part_aliases and common_url_parts for
    // errors.
    String url_part_errors = HtURLCodec::instance()->ErrMsg();

    if (url_part_errors.length() != 0)
      reportError(form("Invalid url_part_aliases or common_url_parts: %s",
                       url_part_errors.get()));


    // We may need these through the methods we call
    if (alt_work_area != 0)
    {
	String	configValue;

	configValue = config->Find("word_db");
	if (configValue.length() != 0)
	{
	    configValue << ".work";
	   config->Add("word_db", configValue);
	}

	configValue = config->Find("doc_db");
	if (configValue.length() != 0)
	{
	    configValue << ".work";
	   config->Add("doc_db", configValue);
	}

	configValue = config->Find("doc_index");
	if (configValue.length() != 0)
	{
	    configValue << ".work";
	   config->Add("doc_index", configValue);
	}

	configValue = config->Find("doc_excerpt");
	if (configValue.length() != 0)
	{
	    configValue << ".work";
	   config->Add("doc_excerpt", configValue);
	}
    }

    if (do_docs)
      {
	const String doc_list = config->Find("doc_list");
	unlink(doc_list);
	DocumentDB docs;
	if (docs.Read(config->Find("doc_db"), config->Find("doc_index"), 
		      config->Find("doc_excerpt")) == OK)
	  {
	    docs.DumpDB(doc_list, verbose);
	    docs.Close();
	  }
      }
    if (do_words)
      {

	// Initialize htword
	WordContext::Initialize(*config);

	const String word_dump = config->Find("word_dump");
	unlink(word_dump);
	HtWordList words(*config);
	if(words.Open(config->Find("word_db"), O_RDONLY) == OK) {
	  words.Dump(word_dump);
	  words.Close();
	}
      }

    return 0;
}


//*****************************************************************************
// void usage()
//   Display program usage information
//
void usage()
{
    cout << "usage: htdump [-v][-d][-w][-a][-c configfile]\n";
    cout << "This program is part of ht://Dig " << VERSION << "\n\n";
    cout << "Options:\n";
    cout << "\t-v\tVerbose mode.  This increases the verbosity of the\n";
    cout << "\t\tprogram.  Using more than 2 is probably only useful\n";
    cout << "\t\tfor debugging purposes.  The default verbose mode\n";
    cout << "\t\tgives a progress on what it is doing and where it is.\n\n";
    cout << "\t-d\tDo NOT dump the document database.\n\n";
    cout << "\t-w\tDo NOT dump the word database.\n\n";
    cout << "\t-a\tUse alternate work files.\n";
    cout << "\t\tTells htdump to append .work to the database files \n";
    cout << "\t\tallowing it to operate on a second set of databases.\n";
    cout << "\t-c configfile\n";
    cout << "\t\tUse the specified configuration file instead on the\n";
    cout << "\t\tdefault.\n\n";
    exit(0);
}


//*****************************************************************************
// Report an error and die
//
void reportError(char *msg)
{
    cout << "htdump: " << msg << "\n\n";
    exit(1);
}
